system/lib/pthread/emscripten_futex_wake.c

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2021 The Emscripten Authors.  All rights reserved.
+ * Emscripten is available under two separate licenses, the MIT license and the
+ * University of Illinois/NCSA Open Source License.  Both these licenses can be
+ * found in the LICENSE file.
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <limits.h>
+#include <atomic.h>
+#include <emscripten/threading.h>
+
+int _emscripten_thread_supports_atomics_wait(void);
+
+// Stores the memory address that the main thread is waiting on, if any. If
+// the main thread is waiting, we wake it up before waking up any workers.
+void* _emscripten_main_thread_futex;
+
+// Returns the number of threads (>= 0) woken up, or the value -EINVAL on error.
+// Pass count == INT_MAX to wake up all threads.
+int emscripten_futex_wake(volatile void *addr, int count) {
+  if (!addr || (((intptr_t)addr) & 3) != 0 || count < 0) {
+    return -EINVAL;
+  }
+  if (count == 0) {
+    return 0;
+  }
+
+  // See if main thread is waiting on this address? If so, wake it up by resetting its wake location to zero.
+  // Note that this is not a fair procedure, since we always wake main thread first before any workers, so
+  // this scheme does not adhere to real queue-based waiting.
+  int main_thread_woken = 0;
+  if (a_cas_p(&_emscripten_main_thread_futex, (void*)addr, 0) == addr) {
+    // We only use __emscripten_main_thread_futex on the main browser thread,
+    // where we cannot block while we wait. Therefore we should only see it set
+    // from other threads (that should always support waiting), and not on the
+    // main thread itself. In other words, the main thread must never try to
+    // wake itself up!
+    assert(_emscripten_thread_supports_atomics_wait());
```

Comment:
At some point we might want to audit our code to try to make it work correctly from audio worklets, which also can't block.

Reply:
Yeah.. this mechanism assumes only one (busy-looping) main thread I think, so it would need to be revisited.

Reply:
I guess I came across the audioworklet issue:
SimpleSequencer.js:1 Uncaught RuntimeError: Aborted(Assertion failed: _emscripten_thread_supports_atomics_wait(), at: ```
/home/jonathan/emsdk/upstream/emscripten/system/lib/pthread/emscripten_futex_wake.c,40,emscripten_futex_wake)
    at abort (SimpleSequencer.js:1:23555)
    at ___assert_fail (SimpleSequencer.js:1:92185)
    at emscripten_futex_wake (SimpleSequencer.wasm:0x5fcc66)
    at __wake (SimpleSequencer.wasm:0x5fc80b)
    at __pthread_mutex_unlock (SimpleSequencer.wasm:0x6041a7)
    at sys_unlock (SimpleSequencer.wasm:0x4b4743)
    at ofxPd::audioOut(float*, int, int) (SimpleSequencer.wasm:0x43abb3)
    at ofxOfelia::audioOut(ofSoundBuffer&) (SimpleSequencer.wasm:0x691b3)
    at std::__2::__function::__func<std::__2::__bind<void (ofBaseSoundInput::*)(ofSoundBuffer&), ofBaseSoundInput*&, std::__2::placeholders::__ph<1> const&>, std::__2::allocator<std::__2::__bind<void (ofBaseSoundInput::*)(ofSoundBuffer&), ofBaseSoundInput*&, std::__2::placeholders::__ph<1> const&>>, void (ofSoundBuffer&)>::operator()(ofSoundBuffer&) (SimpleSequencer.wasm:0x63b62)
    at ofxEmscriptenSoundStream::audio_cb(int, int, int, void*) (SimpleSequencer.wasm:0x637e0)
```